cli/resources.rs

```diff
@@ -489,97 +633,61 @@ impl Future for ChildStatus {
 
   fn poll(&mut self) -> Poll<ExitStatus, ErrBox> {
-    let mut table = RESOURCE_TABLE.lock().unwrap();
-    let maybe_repr = table.get_mut(&self.rid);
-    match maybe_repr {
-      Some(Repr::Child(ref mut child)) => child.poll().map_err(ErrBox::from),
-      _ => Err(bad_resource()),
-    }
+    let mut table = NEW_RESOURCE_TABLE.lock().unwrap();
+    let repr = table.get_mut(&self.rid).ok_or_else(bad_resource)?;
+    let child = repr.downcast_mut::<NewChild>().ok_or_else(bad_resource)?;
+    child.0.poll().map_err(ErrBox::from)
   }
 }
 
 pub fn child_status(rid: ResourceId) -> Result<ChildStatus, ErrBox> {
-  let mut table = RESOURCE_TABLE.lock().unwrap();
-  let maybe_repr = table.get_mut(&rid);
-  match maybe_repr {
-    Some(Repr::Child(ref mut _child)) => Ok(ChildStatus { rid }),
-    _ => Err(bad_resource()),
-  }
+  let table = NEW_RESOURCE_TABLE.lock().unwrap();
+  let repr = table.get(&rid).ok_or_else(bad_resource)?;
+  let _child = &repr.downcast_ref::<NewChild>().ok_or_else(bad_resource)?;
+  Ok(ChildStatus { rid })
 }
 
 pub fn get_repl(rid: ResourceId) -> Result<Arc<Mutex<Repl>>, ErrBox> {
-  let mut table = RESOURCE_TABLE.lock().unwrap();
-  let maybe_repr = table.get_mut(&rid);
-  match maybe_repr {
-    Some(Repr::Repl(ref mut r)) => Ok(r.clone()),
-    _ => Err(bad_resource()),
-  }
+  let table = NEW_RESOURCE_TABLE.lock().unwrap();
+  let repr = table.get(&rid).ok_or_else(bad_resource)?;
+  let repl = &repr.downcast_ref::<NewRepl>().ok_or_else(bad_resource)?;
+  Ok(repl.0.clone())
 }
 
 // TODO: revamp this after the following lands:
 // https://github.com/tokio-rs/tokio/pull/785
 pub fn get_file(rid: ResourceId) -> Result<std::fs::File, ErrBox> {
-  let mut table = RESOURCE_TABLE.lock().unwrap();
+  let mut table = NEW_RESOURCE_TABLE.lock().unwrap();
   // We take ownership of File here.
   // It is put back below while still holding the lock.
-  let maybe_repr = table.remove(&rid);
-
-  match maybe_repr {
-    Some(Repr::FsFile(r)) => {
-      // Trait Clone not implemented on tokio::fs::File,
-      // so convert to std File first.
-      let std_file = r.into_std();
-      // Create a copy and immediately put back.
-      // We don't want to block other resource ops.
-      // try_clone() would yield a copy containing the same
-      // underlying fd, so operations on the copy would also
-      // affect the one in resource table, and we don't need
-      // to write back.
-      let maybe_std_file_copy = std_file.try_clone();
-      // Insert the entry back with the same rid.
-      table.insert(rid, Repr::FsFile(tokio_fs::File::from_std(std_file)));
-
-      maybe_std_file_copy.map_err(ErrBox::from)
-    }
-    _ => Err(bad_resource()),
-  }
+  let mut repr = table.remove(&rid).ok_or_else(bad_resource)?;
+  let fs_file = repr.downcast::<NewFsFile>().ok_or_else(bad_resource)?;
+  // Trait Clone not implemented on tokio::fs::File,
+  // so convert to std File first.
+  // TODO:
+  return Err(bad_resource());
```

Comment:
I'm not sure if downcasting will work in this case, we need to take ownership of the file.